core/network/protocols/block_announce_protocol.cpp

```diff
@@ -0,0 +1,337 @@
+      std::shared_ptr<blockchain::BlockStorage> storage,
+      std::shared_ptr<BabeObserver> babe_observer,
+      std::shared_ptr<crypto::Hasher> hasher,
+      std::shared_ptr<PeerManager> peer_manager)
+      : host_(host),
+        stream_engine_(std::move(stream_engine)),
+        block_tree_(std::move(block_tree)),
+        storage_(std::move(storage)),
+        babe_observer_(std::move(babe_observer)),
+        hasher_(std::move(hasher)),
+        peer_manager_(std::move(peer_manager)) {
+    BOOST_ASSERT(stream_engine_ != nullptr);
+    BOOST_ASSERT(block_tree_ != nullptr);
+    BOOST_ASSERT(storage_ != nullptr);
+    BOOST_ASSERT(babe_observer_ != nullptr);
+    BOOST_ASSERT(hasher_ != nullptr);
+    BOOST_ASSERT(peer_manager_ != nullptr);
+    const_cast<Protocol &>(protocol_) =
+        fmt::format(kBlockAnnouncesProtocol.data(), chain_spec.protocolId());
+  }
+
+  bool BlockAnnounceProtocol::start() {
+    host_.setProtocolHandler(protocol_, [wp = weak_from_this()](auto &&stream) {
+      if (auto self = wp.lock()) {
+        if (auto peer_id = stream->remotePeerId()) {
+          SL_TRACE(self->log_,
+                   "Handled {} protocol stream from: {}",
+                   self->protocol_,
+                   peer_id.value().toBase58());
+          self->onIncomingStream(std::forward<decltype(stream)>(stream));
+          return;
+        }
+        self->log_->warn("Handled {} protocol stream from unknown peer",
+                         self->protocol_);
+        stream->reset();
+      }
+    });
+    return true;
+  }
+
+  bool BlockAnnounceProtocol::stop() {
+    return true;
+  }
+
+  outcome::result<Status> BlockAnnounceProtocol::createStatus() const {
+    /// Roles
+    Roles roles;
+    // TODO(xDimon): Need to set actual role of node
+    //  issue: https://github.com/soramitsu/kagome/issues/678
+    roles.flags.full = 1;
+
+    /// Best block info
+    BlockInfo best_block;
+    const auto &last_finalized = block_tree_->getLastFinalized().block_hash;
+    if (auto best_res =
+            block_tree_->getBestContaining(last_finalized, boost::none);
+        best_res.has_value()) {
+      best_block = best_res.value();
+    } else {
+      log_->error("Could not get best block info: {}",
+                  best_res.error().message());
+      return Error::CAN_NOT_CREATE_STATUS;
+    }
+
+    /// Genesis hash
+    BlockHash genesis_hash;
+    if (auto genesis_res = storage_->getGenesisBlockHash();
+        genesis_res.has_value()) {
+      genesis_hash = std::move(genesis_res.value());
+    } else {
+      log_->error("Could not get genesis block hash: {}",
+                  genesis_res.error().message());
+      return Error::CAN_NOT_CREATE_STATUS;
+    }
+
+    return Status{
+        .roles = roles, .best_block = best_block, .genesis_hash = genesis_hash};
+  }
+
+  void BlockAnnounceProtocol::onIncomingStream(std::shared_ptr<Stream> stream) {
+    BOOST_ASSERT(stream->remotePeerId().has_value());
```

Comment:
Shouldn't check at release?

Reply:
No, should not. It cant enter here before Indentify proticol was handled.